Cython/Compiler/ExprNodes.py

```diff
@@ -14796,7 +14798,16 @@ def check_const(self):
         return self.arg.check_const()
 
     def calculate_result_code(self):
-        return "(%s != 0)" % self.arg.result()
+        arg = self.arg.result()
+        if self.arg.type.is_string:
+            # Test for 0-length string with "ptr[0] != '\0'" instead of just "ptr != 0".
+            # If we'd coerce the pointer to a Python 'bytes' / 'str', we'd do the equivalent
+            # but much less efficiently (strlen -> string object -> len>0).
+            # Unicode strings might fail to decode (so it's not entirely equivalent), but the code
+            # really asks if the C string pointer is non-empty, not the decoded string.
+            return f"({arg} != 0 && ({arg})[0] != 0)"
```

Comment:
Ok, I don't think a change this broad is a good idea. It even breaks code like
```cython
cdef char* mem = <char*> malloc(…)
if not mem:
    raise MemoryError
```
by dereferencing a pointer to uninitialised memory, thus returning an arbitrary undefined result. And I'm sure that pattern is widely used. We definitely need to limit the scope, which means that we'll most likely end up with some inconsistencies in `char*` behaviour.

ISTM that the best path forward is to exclude C strings completely and to let them be pointers w.r.t. boolean tests.

Reply:
Yeah I think the memory allocation check will definitely be used all over the place, so I agree that they should probably just be pointers when used as `if not mem` or `if mem`.

`bool(mem)` behaving as if it was converted to a Python object looks like the existing behaviour so should probably be preserved (which i think this does)